### PR TITLE
(Closes #320) nonstandard stop

### LINKED ISSUE #320: Implement missing executable-constructs in F2008 (R213)
The new block-construct and critical-construct introduced in F2008 are not yet implemented.

Related: https://github.com/stfc/fparser/pull/315#discussion_r829795409_

### PR DESCRIPTION
**Pull request by hiker**
Fixes #320. Adds support for extended stop codes to Fortran 2003, and adds full 2008 support for (even more) flexible expressions.

I hope I did the 2008 part correctly. I esp. wasn't certain how to check if the expressions are constant :(

### CODE REVIEW

File: src/fparser/two/utils.py [hunk 1]
```diff
@@ -114,6 +114,11 @@
 # when reading/writing data using unformatted IO.
 _EXTENSIONS += ["open-convert"]
 
+# While non-standard, many compilers support negative numbers, and string
+# operations in stop statements, e.g. `stop -1` or `stop str1//str2`.
+# With this extension, these statements will be allowed.
+_EXTENSIONS += ["extended-stop-format"]
```
Comment (arporter) [inline src/fparser/two/utils.py:120]: Perhaps s/-format/-args/ here (and elsewhere) as "format" makes me think of Fortran FORMAT statements?

File: src/fparser/two/tests/test_fortran2003.py [hunk 2]
```diff
@@ -2110,6 +2110,14 @@ def test_stop_stmt():  # R849
     assert isinstance(obj, tcls), repr(obj)
```
Comment (arporter) [inline src/fparser/two/tests/test_fortran2003.py:2110]: I'm afraid we're following the "you touch it, you fix it" approach here so please add a docstring for this test.

File: src/fparser/two/tests/fortran2008/test_stop_code_r857.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    assert str(result) == string
+
+
+@pytest.mark.usefixtures("f2008_create")
+@pytest.mark.parametrize("string", ["call sub()", "do i", "1, 2, 3"])
+def test_simple_stop_code_errors(string):
+    """Test that invalid stop codes are handled."""
+    with pytest.raises(NoMatchError) as err:
+        Stop_Code(string)
+    assert f"Stop_Code: '{string}'" in str(err.value)
```
Comment (arporter) [inline src/fparser/two/tests/fortran2008/test_stop_code_r857.py:59]: Just to be on the safe side, please could you add a slightly higher-level test where a STOP statement is parsed in the context of e.g. a program. There have been times when I've added tests like this only to find that the new functionality didn't actually work in practise!

File: src/fparser/two/tests/test_fortran2003.py [hunk 4]
```diff
@@ -2110,6 +2110,14 @@ def test_stop_stmt():  # R849
...
     assert str(obj) == "STOP 'hey you'"
 
+    obj = tcls('stop "123"//"456"')
```
Comment (arporter) [inline src/fparser/two/tests/test_fortran2003.py:2113]: Please could you also test when this extension is disabled.

File: src/fparser/two/Fortran2008/stop_code_r857.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+class Stop_Code(Base):  # R857
+    """
+    Fortran2008 rule R867. Changes the allowed stop code type.
+
+    stop-code is scalar-default-char-constant-expr
+                 or scalar-int-constant-expr
+    """
+
+    subclass_names = ["Scalar_Default_Char_Expr", "Scalar_Int_Expr"]
```
Comment (arporter) [inline src/fparser/two/Fortran2008/stop_code_r857.py:49]: Does this still permit the Fortran2003 forms to match? Please add a test.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks very much for this Joerg. Generally all looks good but I'd like a couple more tests to show that parsing STOP statements works in the context of a full parse and that the extension can be switched off. I'd also like some tests to show that the F2003 forms are still accepted by the F2008 parser.
Updated docs build fine (modulo the hundreds of existing warnings) and look good.

File: doc/source/fparser2.rst [hunk 6]
```diff
@@ -374,6 +374,19 @@ This extension is supported by (at least) the Gnu, Intel and Cray compilers
 but is not a part of any Fortran standard. More details can be found at
 https://gcc.gnu.org/onlinedocs/gfortran/CONVERT-specifier.html
 
+Extended arguments for STOP
++++++++++++++++++++++++++++
+
+May compiler support extended arguments for the STOP statement before Fortran 2008.
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/source/fparser2.rst]: Done.

File: src/fparser/two/Fortran2008/stop_code_r857.py (hunk 5 above)
Comment (hiker) [reply src/fparser/two/Fortran2008/stop_code_r857.py:49]: Test added - I hope I did it correctly.

File: src/fparser/two/tests/fortran2008/test_stop_code_r857.py (hunk 3 above)
Comment (hiker) [reply src/fparser/two/tests/fortran2008/test_stop_code_r857.py:63]: Test for full statements added.

File: src/fparser/two/tests/test_fortran2003.py (hunk 2 above)
Comment (hiker) [reply src/fparser/two/tests/test_fortran2003.py:2122]: Ah - that approach :)

File: src/fparser/two/tests/test_fortran2003.py (hunk 4 above)
Comment (hiker) [reply src/fparser/two/tests/test_fortran2003.py]: Done.

File: src/fparser/two/utils.py (hunk 1 above)
Comment (hiker) [reply src/fparser/two/utils.py]: Yes, that sounds better.

**Comment by hiker** [comment]
Ready for next review.

File: src/fparser/two/tests/fortran2008/test_stop_code_r857.py [hunk 7]
```diff
@@ -39,8 +39,12 @@
 
 import pytest
 from fparser.two.utils import NoMatchError
+from fparser.two import Fortran2003, utils
 from fparser.two.Fortran2008 import Stop_Code
 
+from fparser.api import get_reader
+from fparser.two.utils import NoMatchError, walk
```
Comment (arporter) [inline src/fparser/two/tests/fortran2008/test_stop_code_r857.py:46]: Please merge this with the import at L41.

**Review by arporter — APPROVED** [review]
Thanks for extending those tests Joerg. All good now. I've tested it with the PSyclone test suite and all is fine.
There are couple of bits of tidying needed but as that's just an import statement I'll do that.

**Pull request merged**